### PR TITLE
update stock and make payment via Purchase Invoice

### PR DESCRIPTION
**Pull request by saurabh6790**

### CODE REVIEW

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.js [hunk 1]
```diff
@@ -137,6 +162,44 @@ erpnext.accounts.PurchaseInvoice = erpnext.buying.BuyingController.extend({
 
 cur_frm.script_manager.make(erpnext.accounts.PurchaseInvoice);
 
+// Hide Fields
+// ------------
+cur_frm.cscript.hide_fields = function(doc) {
```
Comment (rmehta) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.js]: cur_frm !!! :do_not_litter:

File: erpnext/controllers/accounts_controller.py [hunk 2]
```diff
@@ -54,6 +54,26 @@ def validate(self):
...
+			is_paid = self.get("is_pos") or self.get("is_paid")
+			if cint(is_paid) == 1:
+				if flt(self.paid_amount) == 0:
+					if self.cash_bank_account:
+						frappe.db.set(self, 'paid_amount',
+							flt(flt(self.grand_total) - flt(self.write_off_amount), self.precision("paid_amount")))
+					else:
+						# show message that the amount is not paid
+						frappe.db.set(self,'paid_amount',0)
+						frappe.msgprint(_("Note: Payment Entry will not be created since 'Cash or Bank Account' was not specified"))
```
Comment (rmehta) [inline erpnext/controllers/accounts_controller.py]: throw exception

File: erpnext/controllers/buying_controller.py [hunk 3]
```diff
@@ -274,3 +274,52 @@ def set_qty_as_per_stock_uom(self):
...
+						"actual_qty": flt(d.rejected_qty) * flt(d.conversion_factor),
+						"serial_no": cstr(d.rejected_serial_no).strip(),
+						"incoming_rate": 0.0
+					}))
+
+		self.bk_flush_supp_wh(sl_entries)
+		self.make_sl_entries(sl_entries, allow_negative_stock=allow_negative_stock,
+			via_landed_cost_voucher=via_landed_cost_voucher)
+	
+	def bk_flush_supp_wh(self, sl_entries):
```
Comment (rmehta) [inline erpnext/controllers/buying_controller.py]: no short forms

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 4]
```diff
@@ -59,6 +80,45 @@ def validate(self):
...
+
+	def validate_rejected_warehouse(self):
+		for d in self.get("items"):
+			if flt(d.rejected_qty) and not d.rejected_warehouse:
+				d.rejected_warehouse = self.rejected_warehouse
+				if not d.rejected_warehouse:
+					frappe.throw(_("Row #{0}: Rejected Warehouse is mandatory against rejected Item {1}").format(d.idx, d.item_code))
+
+	# validate accepted and rejected qty
+	def validate_accepted_rejected_qty(self):
```
Comment (rmehta) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: this function is already there in purchase receipt??

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 5]
```diff
@@ -235,6 +295,38 @@ def update_against_document_in_jv(self):
...
+				'target_dt': 'Purchase Order Item',
+				'join_field': 'po_detail',
+				'target_field': 'returned_qty',
+				'target_parent_dt': 'Purchase Order',
+				# 'target_parent_field': 'per_received',
+				# 'target_ref_field': 'qty',
+				'source_field': '-1 * qty',
+				# 'percent_join_field': 'prevdoc_docname',
+				# 'overflow_type': 'receipt',
+				'extra_cond': """ and exists (select name from `tabPurchase Invoice` where name=`tabPurchase Invoice Item`.parent and is_return=1)"""
```
Comment (nabinhait) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: and update_stock=1

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 6]
```diff
@@ -45,6 +47,25 @@ def validate(self):
...
 
+		# validate cash purchase
+		if (self.is_paid == 1):
+			self.validate_cash()
+
+		# validate stock items
+		if (self.update_stock == 1):
+			pc_obj = frappe.get_doc('Purchase Common')
+			pc_obj.validate_for_items(self)
+
```
Comment (nabinhait) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Don't allow Purchase Invoice to be linked with Purchase Receipt

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 7]
```diff
@@ -45,6 +47,25 @@ def validate(self):
...
+			self.validate_cash()
+
+		# validate stock items
+		if (self.update_stock == 1):
+			pc_obj = frappe.get_doc('Purchase Common')
+			pc_obj.validate_for_items(self)
+			
+			self.validate_purchase_return()
+			self.validate_rejected_warehouse()
+			self.validate_accepted_rejected_qty()
```
Comment (nabinhait) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Commonify all the functions which are present in Purchase Receipt

**Comment by nabinhait** [comment]
Changes needed in Landed Cost Voucher, ability to update landed cost to Purchase Invoice (with Update Stock)

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 8]
```diff
@@ -45,6 +47,25 @@ def validate(self):
...
+
+		# sub-contracting
+		# self.validate_for_subcontracting()
+		# self.create_raw_materials_supplied("supplied_items")
+		# self.set_landed_cost_voucher_amount()
+		# self.update_valuation_rate("items")
```
Comment (nabinhait) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Why all these functions are commented? All the functionalities should be covered.

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 9]
```diff
@@ -335,16 +463,17 @@ def make_gl_entries(self):
...
 		stock_items = self.get_stock_items()
+		warehouse_account = get_warehouse_account()
+		
 		for item in self.get("items"):
 			if flt(item.base_net_amount):
 				account_currency = get_account_currency(item.expense_account)
 				gl_entries.append(
 					self.get_gl_dict({
-						"account": item.expense_account,
+						"account": item.expense_account if not self.update_stock else warehouse_account[item.warehouse]["name"],
```
Comment (nabinhait) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: Set warehouse account as expense account if perpetual inventory enabled and update stock checked. Check set_against_expense_account() function.

**Comment by nabinhait** [comment]
merged manually

**Comment by jof2jc** [comment]
@saurabh6790 @nabinhait is this feature merged to master already? I don't see it..

**Comment by nabinhait** [comment]
@jof2jc No, it will be merged in master along with version 7 (expected in the end of June)

**Comment by jof2jc** [comment]
Thanks @nabinhait

**Pull request closed without merging**